_drafts/practical_zephyr_02_dt_basics.md

```diff
@@ -0,0 +1,1468 @@
+
+This is also what we see in the generated `zephyr.dts`: Since `node_a` is referenced by `node_ref`, Zephyr's DTS generator has inserted the property `phandle` for `node_a` in our devicetree. To see what happens to the reference within this `phandle` property, we need to jump back to the syntax section in the [DTSpec](https://www.devicetree.org/specifications/), where we find the following:
+
+> "In a cell array, a reference to another node will be expanded to that node's phandle."
+
+This means that the references `&{/node_a}` and `&label_a` in our properties `phandle-by-path` and `phandle-by-label` are essentially expanded to `node_a`'s `phandle` _0x1c_. Thus, **the reference is equivalent to its phandle**. Zephyr's documentation is right to refer to `&{/node_a}` and `&label_a` as "`phandle`s". Could we also define `node_a`'s `phandle` property by ourselves? Let's find out:
+
+```dts
+/ {
+  label_a: node_a {
+    phandle = <0xC0FFEE>;
+  };
+  node_refs {
+    phandle-by-path = <&{/node_a}>;
+    phandle-by-label = <&label_a>;
+  };
+};
+```
+
+After executing `west build` with the previous parameters again, we indeed end up with the value `0xc0ffee` for `node_a`'s `phandle` property:
+
+`build/zephyr/zephyr.dts`
+
+```dts
+/dts-v1/;
+
+/ {
+  /* Possibly lots of other nodes ... */
+  label_a: node_a {
+    phandle = < 0xc0ffee >;
+  };
+  node_refs {
+    phandle-by-path = < &{/node_a} >;
+    phandle-by-label = < &label_a >;
+  };
+};
+```
+
+What's the use of a `phandle`? A single `phandle` can be useful where a 1:1 relation between nodes in the devicetree is required. E.g., the following is a snippet from the nRF52840 DTS include file, which contains a software PWM node with a configurable generator, which by default refers to `timer2`:
+
+`zephyr/dts/arm/nordic/nrf52840.dtsi`
+
+```dts
+/ {
+  /* ... */
+  sw_pwm: sw-pwm {
+    compatible = "nordic,nrf-sw-pwm";
+    status = "disabled";
+    generator = <&timer2>;
+    clock-prescaler = <0>;
+    #pwm-cells = <3>;
+  };
+};
+```
+
+We've seen the type of the _phandle_ property, but what about the types of the properties `phandle-by-path` and `phandle-by-label`? Knowing that a `phandle` is expanded to a *cell* we might guess that `phandle-by-path` and `phandle-by-label` are of type `int` or `array`: Assuming that `node_a`'s _phandle_ has the value 0xc0ffee, both `<&{/node_a}>` and `<&label_a>` essentially expand to `<0xc0ffee>`, which could either be a single `int` or an `array` of size `1` containing a 32-bit value.
+
+In Zephyr, however, an array containing a **single** node reference has its own type **`phandle`**.
+
+
+### `path`, `phandles` and `phandle-array`
+
+In addition to the `phandle` type, three more types are available in Zephyr when dealing with `phandle`s and references:
+
+| Zephyr type     | DTSpec equivalent      | Syntax                                   | Example                                |
```

Comment:
OK, now I got rid of all the `<type>` notations since the primitive types from the DTSpec are no longer mentioned as such in the article. Hope it makes it less theoretical :).